src/nn/ema.rs

```diff
@@ -0,0 +1,117 @@
+use super::tensor_collection::{
+    RecursiveWalker, TensorCollection, TensorOptions, TensorVisitor, ViewTensorMut, ViewTensorRef,
+};
+
+use crate::{shapes::*, tensor::*, tensor_ops::ema::EmaKernel};
+
+use std::{string::String, vec::Vec};
+
+struct ModelEMAOp<E> {
+    decay: E,
+}
+impl<E: Dtype, D: EmaKernel<E>> TensorVisitor<E, D> for ModelEMAOp<E> {
+    type Viewer = (ViewTensorMut, ViewTensorRef);
+    type Err = D::Err;
+
+    fn visit<S: Shape>(
+        &mut self,
+        _: String,
+        opts: TensorOptions<S, E, D>,
+        (dst, src): (&mut Tensor<S, E, D>, &Tensor<S, E, D>),
+    ) -> Result<(), Self::Err> {
+        if opts.do_gradient_update {
+            dst.try_ema_assign(src, self.decay)?;
+        }
+        Ok(())
+    }
+}
+
+/// Performs model exponential moving average on two modules.
+pub trait ModelEMA<E: Dtype, D: EmaKernel<E>>: TensorCollection<E, D> {
+    /// Does `self = self * decay + other * (1 - decay), using
+    /// [crate::tensor_ops::ema] on parameters.
+    ///
+    /// **Only updates trainable parameters**. For example, batch normalization
+    /// running parameters are not updated.
+    fn ema_assign(&mut self, other: &Self, decay: E) {
```

Comment:
TODO come up with a better name than this? Is `ema` fine? Do we need something to imply that it is in place?